jetty-security/src/main/java/org/eclipse/jetty/security/DynamicAuthenticator.java

```diff
@@ -0,0 +1,282 @@
+
+import org.eclipse.jetty.http.pathmap.MappedResource;
+import org.eclipse.jetty.http.pathmap.PathMappings;
+import org.eclipse.jetty.http.pathmap.ServletPathSpec;
+import org.eclipse.jetty.server.Authentication;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.URIUtil;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.security.Constraint;
+
+/**
+ * Used as a mapping of {@link org.eclipse.jetty.http.pathmap.PathSpec} to {@link Authenticator}.
+ *
+ * This allows you to register different authentication mechanisms for different paths.
+ */
+public class DynamicAuthenticator extends ContainerLifeCycle implements Authenticator
+{
+    private static final Logger LOG = Log.getLogger(DynamicAuthenticator.class);
+
+    private final PathMappings<AuthenticatorEntry> _mappings = new PathMappings<>();
+
+    public void addMapping(String pathSpec, Authenticator authenticator)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator));
+    }
+
+    public void addMapping(String pathSpec, Authenticator authenticator, LoginService loginService)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator, loginService));
+    }
+
+    public void addMapping(String pathSpec, Authenticator authenticator, LoginService loginService, IdentityService identityService)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator, loginService, identityService));
+    }
+
+    private void addMapping(String pathSpec, AuthenticatorEntry entry)
+    {
+        _mappings.put(new ServletPathSpec(pathSpec), entry);
+        addBean(entry);
+    }
+
+    @Override
+    public void setConfiguration(AuthConfiguration configuration)
+    {
+        for (MappedResource<AuthenticatorEntry> mapping : _mappings)
+        {
+            AuthenticatorEntry entry = mapping.getResource();
+            DynamicConfiguration config = new DynamicConfiguration(configuration, entry);
+            entry.getAuthenticator().setConfiguration(config);
+        }
+    }
+
+    @Override
+    public String getAuthMethod()
+    {
+        return Constraint.__DYNAMIC_AUTH;
+    }
+
+    @Override
+    public void prepareRequest(ServletRequest request)
+    {
+        Authenticator authenticator = getAuthenticator(request);
+        if (authenticator == null)
+            return;
+
+        authenticator.prepareRequest(request);
+    }
+
+    @Override
+    public Authentication validateRequest(ServletRequest request, ServletResponse response, boolean mandatory) throws ServerAuthException
+    {
+        final Request baseRequest = Objects.requireNonNull(Request.getBaseRequest(request));
+        final Response baseResponse = baseRequest.getResponse();
+
+        Authenticator authenticator = getAuthenticator(request);
```

Comment:
Don't we need to be able to validate with all known authenticators regardless of path?

Consider a webapp for which we will accept either FORM or OPEN_ID authentication.  I think this needs to be configured so that FORM auth is the authenticator for all pages except one that does the open ID authentication.   Then when an unauthenticated client hits any page they are redirected to the FORM login page.  There they can either put in their credentials and proceed OR click the link for OPEN_ID auth that will take them to the one page that runs the open id conversation.

After either authentication paths, they should be able to see any secured page, regardless of the auth method used, so their requests need to be authenticated against all authenticators... or at least all matching authenticators in the case of overlapping patterns.

Reply:
I don't see how the proposed mechanism will deal with the user deciding to opt for, say, openId authentication after having got to the login page for form authentication. In this case, the user will have hit a protected url that requires authentication; the form authenticator has been selected (as the default) and remembered the url the user was trying to access, and redirected to the login page. If the user then clicked on a "Use OpenId Authentication"  button, while the server could select the OpenIdAuthenticator, it wouldn't know which url it needed to redirect the user back to after a successful authentication. AFAIK the FormAuthenticator and OpenIdAuthenticator don't share common session attributes to store that info.

I think the proposed path:authenticator mapping needs more work to clarify. At the moment it feels too tied up with the security-constraint mappings. The security-constraint path mappings should be agnostic of the authentication alternatives.  Another way to indicate which authenticator to select could be via a new, well-known request parameter. However, the bigger problem is that the authenticator alternatives all need to cooperate at a deeper level as outlined in the use case above.